perfkitbenchmarker/publisher.py

```diff
@@ -538,6 +548,87 @@ def PublishSamples(self, samples):
       vm_util.IssueRetryableCommand(copy_cmd)
 
 
+class ElasticsearchPublisher(SamplePublisher):
+  """Publish samples to an Elasticsearch server. Index and document type
+  will be created if they do not exist.
+
+  Attributes:
+    es_uri: String. i.e. http://IP:port. Default "http://localhost:9200"
+    es_index: String. Default "perfkit"
+    es_type: String. Default "result"
+  """
+  def __init__(self, es_uri=None, es_index=None, es_type=None):
+    self.es_uri = es_uri
+    self.es_index = es_index.lower()
+    self.es_type = es_type
+    self.mapping = {
+        "mappings": {
+            "result": {
+                "numeric_detection": True,
+                "properties": {
+                    "timestamp": {
+                        "type": "date",
+                        "format": "yyyy-MM-dd HH:mm:ss.SSSSSS"
+                    },
+                    "value": {
+                        "type": "double"
+                    }
+                },
+                "dynamic_templates": [{
+                    "strings": {
+                        "match_mapping_type": "string",
+                        "mapping": {
+                            "type": "string",
+                            "fields": {
+                                "raw": {
+                                    "type": "string",
+                                    "index": "not_analyzed"
+                                }
+                            }
+                        }
+                    }
+                }]
+            }
+        }
+    }
+
+  def PublishSamples(self, samples):
+    """Publish samples to Elasticsearch service"""
+    es = Elasticsearch([self.es_uri])
+    if not es.indices.exists(index=self.es_index):
+      es.indices.create(index=self.es_index, body=self.mapping)
+      logging.info('Create index %s and default mappings', self.es_index)
+    for sample in samples:
+      sample = sample.copy()
+      # Make timestamp understandable by ES and human.
+      sample['timestamp'] = self._HumanReadableTimestamp(sample['timestamp'])
+      # Keys cannot have dots for ES
+      sample = self._deDotKeys(sample)
+      # Add sample to the "perfkit index" of "result type" and using sample_uri
+      # as each ES's document's unique _id
+      es.create(index=self.es_index, doc_type=self.es_type,
+                id=sample['sample_uri'], body=json.dumps(sample))
+
+  def _HumanReadableTimestamp(self, epoch_us):
+    """Convert the floating epoch timestamp in micro seconds epoch_us to
+    yyyy-MM-dd HH:mm:ss.SSSSSS in string
+    """
+    ts = time.strftime('%Y-%m-%d %H:%M:%S', time.gmtime(epoch_us))
+    num_dec = ("%.6f" % (epoch_us - int(epoch_us))).split('.')[1]
+    new_ts = '%s.%s' % (ts, num_dec)
+    return new_ts
+
+  def _deDotKeys(self, res):
+    """Recursively replace dot with underscore in all keys in a """
```

Comment:
Looks like an incomplete doc string. Assuming you meant "all keys in a dictionary"?